File: scripts/beam_pipelines/streaming_history_transform.py
#!/usr/bin/env python3
"""
Spotify Data Pipeline - Apache Beam Transformation for Streaming History

This script uses Apache Beam to transform streaming history data for analytics.
It performs the following transformations:
- Calculates play duration statistics by user, track, and artist
- Computes completion rates for tracks
- Identifies listening patterns by time of day
- Groups plays by platform and country
"""

import argparse
import datetime
import json
import logging
import os
import sys
from typing import Dict, List, Tuple, Any

import apache_beam as beam
from apache_beam.io import ReadFromText, WriteToText
from apache_beam.io.jdbc import ReadFromJdbc, WriteToJdbc
from apache_beam.options.pipeline_options import PipelineOptions, SetupOptions
from apache_beam.transforms import PTransform, DoFn, GroupByKey, CombinePerKey
from apache_beam.transforms.core import ParDo
from dotenv import load_dotenv

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)

# Load environment variables
load_dotenv()

# Database connection parameters
DB_HOST = os.getenv("DB_HOST", "postgres")
DB_PORT = os.getenv("DB_PORT", "5432")
DB_NAME = os.getenv("DB_NAME", "spotify_raw")
DB_USER = os.getenv("DB_USER", "postgres")
DB_PASSWORD = os.getenv("DB_PASSWORD", "postgres")
DB_ANALYTICS = os.getenv("DB_ANALYTICS_NAME", "spotify_analytics")

# JDBC connection strings
JDBC_RAW_URL = f"jdbc:postgresql://{DB_HOST}:{DB_PORT}/{DB_NAME}"
JDBC_ANALYTICS_URL = f"jdbc:postgresql://{DB_HOST}:{DB_PORT}/{DB_ANALYTICS}"
JDBC_DRIVER = "org.postgresql.Driver"

# SQL queries
STREAMING_HISTORY_QUERY = """
SELECT 
    event_id, user_id, track_id, timestamp, ms_played, 
    reason_start, reason_end, shuffle, skipped, platform, country
FROM raw.streaming_history
WHERE DATE(timestamp) = ?
"""

# Beam transforms
class ExtractStreamingData(DoFn):
    """Extract and parse streaming history data."""
    
    def process(self, element):
        """Process each streaming history record."""
        try:
            # Convert timestamp to datetime
            if isinstance(element['timestamp'], str):
                element['timestamp'] = datetime.datetime.fromisoformat(
                    element['timestamp'].replace('Z', '+00:00')
                )
            
            # Calculate hour of day for time-based analysis
            hour_of_day = element['timestamp'].hour
            
            # Determine time of day category
            if 5 <= hour_of_day < 12:
                time_of_day = 'morning'
            elif 12 <= hour_of_day < 17:
                time_of_day = 'afternoon'
            elif 17 <= hour_of_day < 22:
                time_of_day = 'evening'
            else:
                time_of_day = 'night'
            
            # Determine if weekend
            is_weekend = element['timestamp'].weekday() >= 5
            
            # Add derived fields
            element['time_of_day'] = time_of_day
            element['is_weekend'] = is_weekend
            element['date'] = element['timestamp'].date().isoformat()
            
            # Calculate completion ratio if we have duration data
            if 'ms_played' in element and element.get('ms_played') is not None:
                # Note: We would need track duration to calculate actual completion ratio
                # This is a placeholder estimation
                completion_ratio = min(1.0, element['ms_played'] / 180000)  # Assume 3 min avg
                element['completion_ratio'] = completion_ratio
            
            yield element
        
        except Exception as e:
            logger.error(f"Error processing record: {e}")
            # Skip the record
            return


class CalculateUserListeningStats(DoFn):
    """Calculate listening statistics by user and date."""
    
    def process(self, element):
        """Process group of records for a user on a date."""
        (key, records) = element
        user_id, date = key
        
        # Initialize counters
        total_tracks = 0
        total_time_ms = 0
        track_durations = []
        artists = {}
        genres = {}
        time_of_day_counts = {
            'morning': 0,
            'afternoon': 0,
            'evening': 0,
            'night': 0
        }
        
        # Process records
        for record in records:
            total_tracks += 1
            
            # Accumulate play time
            if 'ms_played' in record and record['ms_played'] is not None:
                total_time_ms += record['ms_played']
                track_durations.append(record['ms_played'])
            
            # Count by time of day
            if 'time_of_day' in record:
                time_of_day_counts[record['time_of_day']] += 1
            
            # Track artist counts (for favorite artist calculation)
            if 'artist_id' in record:
                artists[record['artist_id']] = artists.get(record['artist_id'], 0) + 1
            
            # Track genre counts (would come from artist or track metadata)
            if 'genre' in record:
                genres[record['genre']] = genres.get(record['genre'], 0) + 1
        
        # Calculate averages and favorites
        avg_track_duration_ms = sum(track_durations) / len(track_durations) if track_durations else 0
        favorite_artist_id = max(artists.items(), key=lambda x: x[1])[0] if artists else None
        favorite_genre = max(genres.items(), key=lambda x: x[1])[0] if genres else None
        
        # Create result record
        result = {
            'date': date,
            'user_id': user_id,
            'total_tracks': total_tracks,
            'total_time_ms': total_time_ms,
            'avg_track_duration_ms': avg_track_duration_ms,
            'favorite_artist_id': favorite_artist_id,
            'favorite_genre': favorite_genre,
            'morning_tracks': time_of_day_counts['morning'],
            'afternoon_tracks': time_of_day_counts['afternoon'],
            'evening_tracks': time_of_day_counts['evening'],
            'night_tracks': time_of_day_counts['night']
        }
        
        yield result


class CalculateTrackPopularity(DoFn):
    """Calculate popularity statistics by track and date."""
    
    def process(self, element):
        """Process group of records for a track on a date."""
        (key, records) = element
        track_id, date = key
        
        # Initialize counters
        total_streams = 0
        unique_listeners = set()
        completion_rates = []
        skips = 0
        
        # Process records
        for record in records:
            total_streams += 1
            unique_listeners.add(record['user_id'])
            
            # Track completion rates
            if 'completion_ratio' in record:
                completion_rates.append(record['completion_ratio'])
            
            # Track skips
            if record.get('skipped'):
                skips += 1
        
        # Calculate statistics
        avg_completion_rate = sum(completion_rates) / len(completion_rates) if completion_rates else 0
        skip_rate = skips / total_streams if total_streams > 0 else 0
        
        # Create result record
        result = {
            'date': date,
            'track_id': track_id,
            'total_streams': total_streams,
            'unique_listeners': len(unique_listeners),
            'avg_completion_rate': avg_completion_rate,
            'skip_rate': skip_rate
        }
        
        yield result


def run_pipeline(date_str, output_path=None):
    """Run the Apache Beam pipeline for the specified date."""
    # Pipeline options
    options = PipelineOptions([
        '--runner=DirectRunner',
        f'--temp_location=/opt/beam/temp'
    ])
    options.view_as(SetupOptions).save_main_session = True
    
    # Parse the date
    try:
        date = datetime.datetime.strptime(date_str, '%Y-%m-%d').date()
    except ValueError:
        logger.error(f"Invalid date format: {date_str}. Use YYYY-MM-DD.")
        return
    
    # Create pipeline
    with beam.Pipeline(options=options) as pipeline:
        # Read streaming history from database
        streaming_data = (
            pipeline
            | "ReadFromJDBC" >> ReadFromJdbc(
                table_name="raw.streaming_history",
                driver_class_name=JDBC_DRIVER,
                jdbc_url=JDBC_RAW_URL,
                username=DB_USER,
                password=DB_PASSWORD,
                query=STREAMING_HISTORY_QUERY,
                statement_parameters=[date_str]
            )
            | "ExtractStreamingData" >> ParDo(ExtractStreamingData())
        )
        
        # Calculate daily listening stats by user
        (
            streaming_data
            | "KeyByUserAndDate" >> beam.Map(lambda x: ((x['user_id'], x['date']), x))
            | "GroupByUserAndDate" >> GroupByKey()
            | "CalculateUserListeningStats" >> ParDo(CalculateUserListeningStats())
            | "WriteDailyListeningStats" >> WriteToJdbc(
                table_name="analytics.daily_listening_stats",
                driver_class_name=JDBC_DRIVER,
                jdbc_url=JDBC_ANALYTICS_URL,
                username=DB_USER,
                password=DB_PASSWORD
            )
        )
        
        # Calculate track popularity
        (
            streaming_data
            | "KeyByTrackAndDate" >> beam.Map(lambda x: ((x['track_id'], x['date']), x))
            | "GroupByTrackAndDate" >> GroupByKey()
            | "CalculateTrackPopularity" >> ParDo(CalculateTrackPopularity())
            | "WriteTrackPopularity" >> WriteToJdbc(
                table_name="analytics.track_popularity",
                driver_class_name=JDBC_DRIVER,
                jdbc_url=JDBC_ANALYTICS_URL,
                username=DB_USER,
                password=DB_PASSWORD
            )
        )
        
        # If output path is specified, also write to file
        if output_path:
            (
                streaming_data
                | "FormatForOutput" >> beam.Map(lambda x: json.dumps(x))
                | "WriteToText" >> WriteToText(output_path)
            )


def main():
    """Main function to run the Apache Beam pipeline."""
    parser = argparse.ArgumentParser(description='Transform streaming history data using Apache Beam')
    parser.add_argument('--date', required=True, help='Date to process in YYYY-MM-DD format')
    parser.add_argument('--output', help='Optional output file path')
    args = parser.parse_args()
    
    # Run the pipeline
    run_pipeline(args.date, args.output)


if __name__ == "__main__":
    main() 

File: src/scio/scio_pipelines/src/main/scala/com/spotify/pipeline/transforms/StreamingHistoryTransform.scala
package com.spotify.pipeline.transforms

import com.spotify.scio._
import com.spotify.scio.jdbc._
import com.spotify.scio.values.SCollection
import com.typesafe.config.ConfigFactory
import org.apache.beam.sdk.options.PipelineOptionsFactory
import org.slf4j.LoggerFactory

import java.sql.Timestamp
import java.time.{DayOfWeek, LocalDateTime, ZoneId, ZonedDateTime}
import java.time.format.DateTimeFormatter
import scala.util.{Failure, Success, Try}

/**
 * StreamingHistoryTransform - A Scio pipeline for transforming Spotify streaming history data
 * 
 * This pipeline:
 * 1. Reads streaming history from PostgreSQL
 * 2. Extracts time-based patterns and calculates metrics
 * 3. Computes user listening stats and track popularity
 * 4. Writes results back to analytics tables
 */
object StreamingHistoryTransform {
  // Logger
  private val logger = LoggerFactory.getLogger(this.getClass)
  
  // Load configuration
  private val config = ConfigFactory.load()
  private val dbConfig = config.getConfig("spotify.pipeline.database")
  private val rawDbConfig = dbConfig.getConfig("raw")
  private val analyticsDbConfig = dbConfig.getConfig("analytics")
  private val beamConfig = config.getConfig("spotify.pipeline.beam")
  private val streamingHistoryConfig = config.getConfig("spotify.pipeline.streaming_history")
  
  // Case classes for data models
  case class StreamingEvent(
    eventId: String,
    userId: String,
    trackId: String,
    timestamp: Timestamp,
    msPlayed: Option[Int],
    reasonStart: Option[String],
    reasonEnd: Option[String],
    shuffle: Option[Boolean],
    skipped: Option[Boolean],
    platform: Option[String],
    country: Option[String]
  )

  case class EnrichedStreamingEvent(
    eventId: String,
    userId: String,
    trackId: String,
    timestamp: Timestamp,
    date: String,
    msPlayed: Option[Int],
    reasonStart: Option[String],
    reasonEnd: Option[String],
    shuffle: Option[Boolean],
    skipped: Option[Boolean],
    platform: Option[String],
    country: Option[String],
    timeOfDay: String,
    isWeekend: Boolean,
    completionRatio: Option[Double]
  )
  
  case class UserListeningStats(
    date: String,
    userId: String,
    totalTracks: Int,
    totalTimeMs: Long,
    avgTrackDurationMs: Double,
    favoriteArtistId: Option[String],
    favoriteGenre: Option[String],
    morningTracks: Int,
    afternoonTracks: Int,
    eveningTracks: Int,
    nightTracks: Int
  )
  
  case class TrackPopularity(
    date: String,
    trackId: String,
    totalStreams: Int,
    uniqueListeners: Int,
    avgCompletionRate: Double,
    skipRate: Double
  )
  
  def main(args: Array[String]): Unit = {
    // Parse date argument
    val date = if (args.length > 0) args(0) else {
      val today = java.time.LocalDate.now()
      today.format(DateTimeFormatter.ISO_LOCAL_DATE)
    }
    
    logger.info(s"Starting StreamingHistoryTransform for date: $date")
    
    // Set up pipeline options
    val options = PipelineOptionsFactory.create()
    options.setJobName(s"spotify-streaming-history-transform-$date")
    
    // Create ScioContext
    val sc = ScioContext(options)
    
    // Read streaming history from PostgreSQL
    val rawJdbcOptions = JdbcConnectionOptions(
      username = rawDbConfig.getString("user"),
      password = rawDbConfig.getString("password"),
      driverClass = rawDbConfig.getString("driver")
    )
    
    val analyticsJdbcOptions = JdbcConnectionOptions(
      username = analyticsDbConfig.getString("user"),
      password = analyticsDbConfig.getString("password"),
      driverClass = analyticsDbConfig.getString("driver")
    )
    
    val streamingEvents: SCollection[StreamingEvent] = sc.jdbcSelect(
      rawDbConfig.getString("url"),
      streamingHistoryConfig.getString("sql_query"),
      statementPreparator = statement => { statement.setString(1, date) },
      connectionOptions = rawJdbcOptions
    )(rs => 
      StreamingEvent(
        rs.getString("event_id"),
        rs.getString("user_id"),
        rs.getString("track_id"),
        rs.getTimestamp("timestamp"),
        Option(rs.getObject("ms_played")).map(_.asInstanceOf[Integer].toInt),
        Option(rs.getString("reason_start")),
        Option(rs.getString("reason_end")),
        Option(rs.getObject("shuffle")).map(_.asInstanceOf[Boolean]),
        Option(rs.getObject("skipped")).map(_.asInstanceOf[Boolean]),
        Option(rs.getString("platform")),
        Option(rs.getString("country"))
      )
    )
    
    // Enrich streaming events with derived data
    val enrichedEvents: SCollection[EnrichedStreamingEvent] = streamingEvents.map { event =>
      val localDateTime = event.timestamp.toLocalDateTime
      
      // Determine time of day category
      val hour = localDateTime.getHour
      val timeOfDay = hour match {
        case h if h >= 5 && h < 12 => "morning"
        case h if h >= 12 && h < 17 => "afternoon"
        case h if h >= 17 && h < 22 => "evening"
        case _ => "night"
      }
      
      // Determine if weekend
      val isWeekend = {
        val dayOfWeek = localDateTime.getDayOfWeek
        dayOfWeek == DayOfWeek.SATURDAY || dayOfWeek == DayOfWeek.SUNDAY
      }
      
      // Calculate completion ratio (estimated)
      val completionRatio = event.msPlayed.map { ms =>
        math.min(1.0, ms.toDouble / 180000.0) // Assume 3 min average track length
      }
      
      // Format date in ISO format (YYYY-MM-DD)
      val date = localDateTime.toLocalDate.format(DateTimeFormatter.ISO_LOCAL_DATE)
      
      EnrichedStreamingEvent(
        event.eventId,
        event.userId,
        event.trackId,
        event.timestamp,
        date,
        event.msPlayed,
        event.reasonStart,
        event.reasonEnd,
        event.shuffle,
        event.skipped,
        event.platform,
        event.country,
        timeOfDay,
        isWeekend,
        completionRatio
      )
    }
    
    // Calculate user listening stats
    val userListeningStats: SCollection[UserListeningStats] = enrichedEvents
      .keyBy(e => (e.date, e.userId))
      .groupByKey
      .map { case ((date, userId), events) =>
        // Initialize counters
        var totalTracks = 0
        var totalTimeMs = 0L
        val trackDurations = scala.collection.mutable.ArrayBuffer.empty[Int]
        val artists = scala.collection.mutable.Map.empty[String, Int]
        val genres = scala.collection.mutable.Map.empty[String, Int]
        val timeOfDayCounts = scala.collection.mutable.Map(
          "morning" -> 0,
          "afternoon" -> 0,
          "evening" -> 0,
          "night" -> 0
        )
        
        // Process events
        events.foreach { event =>
          totalTracks += 1
          
          // Accumulate play time
          event.msPlayed.foreach { ms =>
            totalTimeMs += ms
            trackDurations += ms
          }
          
          // Count by time of day
          timeOfDayCounts(event.timeOfDay) += 1
          
          // TODO: In a real implementation, we would join with artist data
          // For now, we'll leave these as None
        }
        
        // Calculate averages
        val avgTrackDurationMs = if (trackDurations.nonEmpty) {
          trackDurations.sum.toDouble / trackDurations.size
        } else {
          0.0
        }
        
        // Determine favorites (would be derived from actual data)
        val favoriteArtistId = None // In reality, would be: artists.maxBy(_._2)._1
        val favoriteGenre = None // In reality, would be: genres.maxBy(_._2)._1
        
        UserListeningStats(
          date,
          userId,
          totalTracks,
          totalTimeMs,
          avgTrackDurationMs,
          favoriteArtistId,
          favoriteGenre,
          timeOfDayCounts("morning"),
          timeOfDayCounts("afternoon"),
          timeOfDayCounts("evening"),
          timeOfDayCounts("night")
        )
      }
    
    // Calculate track popularity
    val trackPopularity: SCollection[TrackPopularity] = enrichedEvents
      .keyBy(e => (e.date, e.trackId))
      .groupByKey
      .map { case ((date, trackId), events) =>
        // Initialize counters
        var totalStreams = 0
        val uniqueListeners = scala.collection.mutable.Set.empty[String]
        val completionRates = scala.collection.mutable.ArrayBuffer.empty[Double]
        var skips = 0
        
        // Process events
        events.foreach { event =>
          totalStreams += 1
          uniqueListeners += event.userId
          
          // Track completion rates
          event.completionRatio.foreach(completionRates += _)
          
          // Track skips
          if (event.skipped.getOrElse(false)) {
            skips += 1
          }
        }
        
        // Calculate statistics
        val avgCompletionRate = if (completionRates.nonEmpty) {
          completionRates.sum / completionRates.size
        } else {
          0.0
        }
        
        val skipRate = if (totalStreams > 0) {
          skips.toDouble / totalStreams
        } else {
          0.0
        }
        
        TrackPopularity(
          date,
          trackId,
          totalStreams,
          uniqueListeners.size,
          avgCompletionRate,
          skipRate
        )
      }
    
    // Write user listening stats to analytics database
    userListeningStats.saveAsJdbc(
      analyticsDbConfig.getString("url"),
      "analytics.daily_listening_stats",
      connectionOptions = analyticsJdbcOptions
    )
    
    // Write track popularity to analytics database
    trackPopularity.saveAsJdbc(
      analyticsDbConfig.getString("url"),
      "analytics.track_popularity",
      connectionOptions = analyticsJdbcOptions
    )
    
    // Execute the pipeline
    val result = sc.run().waitUntilFinish()
    logger.info(s"Pipeline completed with state: ${result.toString}")
  }
} 